keras_retinanet/bin/train.py

```diff
@@ -287,8 +254,37 @@ def csv_list(string):
 
     return check_args(parser.parse_args(args))
 
+def load_plugins(plugin_path):
+    """
+    Responsible for initialising the plugin manager, setting the plugin directory to search, and loading all available
+    plugins - then activating them.
+
+    :param plugin_path: String/[Str] for plugin paths to check
+    :return: None
+    """
+
+    pm = PluginManagerSingleton.get()
```

Comment:
The elegance of the singleton is that it should be accessible from everywhere. I don't necessarily think that creating this at program time, and passing it everywhere is very elegant. If it's needed in a new location ( quite feasible ), then it becomes a pain to weave this through. The reason I chose the `Singleton` variant is just so that we can't accidentally make a secondary plugin system and have two lots of state.

Are you saying to create a dictionary mapping that lies within the train script? This removes from the advantage of the plugin system. Currently the training implementation needs to know nothing about plugins specific implementation, it allows plugins to self-describe themselves and what they trigger to. If we make a static mapping, it defeats the purpose of this mechanism surely?

Perhaps the naming of this function is a misnomer. It more accurately reflects plugin discovery and initialisation. Where it searches based on defined parameters, and info files, and creates objects of those via `collectPlugins`.

Reply:
> The elegance of the singleton is that it should be accessible from everywhere. I don't necessarily think that creating this at program time, and passing it everywhere is very elegant. If it's needed in a new location ( quite feasible ), then it becomes a pain to weave this through. The reason I chose the Singleton variant is just so that we can't accidentally make a secondary plugin system and have two lots of state.

The flip side is that you get an near invisible dependency on the singleton in potentially many places. That makes it harder to switch to a different plugin system if we ever want to. It's is also not clear to me if the singleton will be sufficient when we want to add a second plugin type (for backbone selection for example).

> Are you saying to create a dictionary mapping that lies within the train script? This removes from the advantage of the plugin system. Currently the training implementation needs to know nothing about plugins specific implementation, it allows plugins to self-describe themselves and what they trigger to. If we make a static mapping, it defeats the purpose of this mechanism surely?

What I mean is that we can build a dictionary of dataset types to plugin objects from the list of plugins found by the plugin manager. That way we can completely hide (and decouple) the system used to find the plugins from the train script.

The added advantage is that a fallback to a hardcoded dictionary is really easy to implement so that yapsy can remain an optional dependency.

Reply:
> The flip side is that you get an near invisible dependency on the singleton in potentially many places. That makes it harder to switch to a different plugin system if we ever want to. It's is also not clear to me if the singleton will be sufficient when we want to add a second plugin type (for backbone selection for example).

Implementing another plugin system would just be a case of either utilising its singleton, or wrapping in our own singleton to make sure only one exists. At the time of choosing the singleton, we weren't requiring it for backbone type, I had considered it. In that case we would just create wrappers for `Dataset` and `Backbone` plugin systems if we wanted to keep them separate.

>What I mean is that we can build a dictionary of dataset types to plugin objects from the list of plugins found by the plugin manager. That way we can completely hide (and decouple) the system used to find the plugins from the train script.
Sounds like more evidence for creating a util wrapper to me? If we created our own util DatasetPluginManger which handled the creation of the plugin system, initialisation, loading, and had some utility funcs like `load_plugins` which returns this dict?

From the sounds of it, that util wrapper is just to act as a middleman between utilisation within train and the implementation of the plugin solution.

Reply:
> From the sounds of it, that util wrapper is just to act as a middleman between utilisation within train and the implementation of the plugin solution.

Yes, with the advantage that it decouples the scripts from the plugin manager and provides an easy minimal interface that can be shared with the other scripts.

Reply:
Sounds good. Proposed API for this? Anything in particular you'd need from this.
I'm thinking:
* Initialisation of Plugin Manager against 'plugins' folder
* Load plugins - Searches for plugins and collects them, creating them.
* getPlugins - For returning this dict of name -> plugin object mappings.

Reply:
I was thinking simply a single function: `load_plugins(paths)`. The one argument would be a list of search paths, the returned value a dictionary mapping `dataset_type` to `plugin_object`.